doc/specs/stdlib_hash_functions.md

```diff
@@ -0,0 +1,1697 @@
+| ifort| NA? | NA? | overflow wrapping |
+| NAG Fortran | -C=none | -C=intovf | -C=none |
+| NEC Fortran | NA? | NA? | overflow wrapping? |
+| NVIDIA Fortran | NA? | NA? | overflow wrapping? |
+
+All of the modules' hash functions take one or two arguments.
+All of them have as their first argument the object to be hashed,
+termed a *key*.
+Most have a second argument, termed a *seed*, that sets the initial
+value of the hash code changing the hash function behavior.
+In particular, inputs that hash to the same hash index with a given
+seed, will often hash to different indexes with a different seed.
+This difference in behavior makes algorithms that use a seed much
+more resistant to denial of service attacks that use the properties
+of a known hash to increase the number of hash table collisions.
+This additional integer must be kept the same for all hashes
+in a given hash table, but can be changed and the objects rehashed
+if collisions are unusually common.
+The *seed* can be either a scalar or a two element array.
+Some of the hash functions have alternatives that allow incremental
+hashing. 
+
+|Algorithm|Seed|Result|
+|---------|----|------|
+|FNV-1|None|32 or 64 bit integer|
+|FNV-1a|None|32 or 64 bit integer|
+|nmhash32 |32 bit scalar integer|32 bit integer|
+|nmhash32x |32 bit scalar integer|32 bit integer|
+|pengyhash |32 bit scalar integer|64 bit integer|
+|Spooky Hash|64 bit two element vector|64 bit two element vector|
+|waterhash|64 bit scalar integer|32 bit integer|
+
+The hash function modules each provide at least five algorithms for
+hash functions: two optimized for small (< 32 `INT8` integer elements)
+keys, and three optimized for large (> 100 `INT8` integer elements)
+keys.
+The core implementation for each algorithm is for keys that are
+vectors of `INT8` integers.
+These core implementations are then used in wrappers for keys
+that are vectors of `INT16`, `INT32` and `INT64` integers, or default
+character strings, in the expectation that inlining will eliminate the
+overhead of transferring the other keys to `INT8` integer vectors.
+
+The `stdlib_32_bit_hash_functions` module provides
+implementations of five hash code algorithms:
+the *FNV_1* and *FNV_1A* variants of Glenn Fowler, 
+Landon Curt Noll, and Kiem-Phong Vo;
+the *nmhash32* and *nmhash32x* of James Z. M. Gao;
+and the *waterhash*  of Tommy Ettinger.
+The detailed implementation of each algorithm is handled in a separate
+submodule: `stdlib_32_bit_fnv_hashes`,
+`stdlib_32_bit_nmhashes`, and `stdlib_32_bit_water_hashes`,
+respectively. The `nmhash32`, `nmhash32x`, and `waterhash` algorithms
+require seeds. The submodules provide separate seed generators
+for each algorithm.
+The module itself
+implements two scalar hash functions, `FIBONACCI_HASH` and
+`UNIVERSAL_MULT_HASH`. 
+It also implements the subroutine, `ODD_RANDOM_INTEGER`, for
+generating seeds for `UNIVERSAL_MULT_HASH`.
+All assume a two's complement sign bit, and no out of
+range checks.
```

Comment:
What do you mean by "no out of range checks", and why is it needed?

Reply:
When  you perform an addition, subtraction, or multiplication the nominal result can be outside the range of integer values represented by the bit model for that integer, i.e., you can have overflow or underflow.  Some compilers have the option of detecting and reporting when overflow or underflow occurs, and stop when it is detected. Many of the hash functions use multiplication or addition, and you don't want them to stop on integer overflow or underflow.

Reply:
ok. Makes sense. Thank you for the explanation.